### PR TITLE
feat(masthead-v2): optimize megamenu tabs

### LINKED ISSUE #10667: [Masthead]: optimize component to lower DOM node count
JIRA: https://jsw.ibm.com/browse/ADCMS-3667

Currently this component has a DOM node count which can be lowered. Solutions to consider are removing duplicate/unnecessary nodes and lazy loading content.

Work has already been completed for the cloud version of the masthead https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/10085.

| Component | Nodes| Comments | Inters |
|--------|--------|--------|--------|
| Masthead composite - sidebar open | 5329 | 1209 | 2349 |
| Masthead composite - topnav open | 2544 | 578 | 1281 |
| Masthead composite/container | 1319 | 323 | 665 |

### PR DESCRIPTION
**Pull request by jkaeser**
### Related Ticket(s)

Resolves #10667 

### Description

Tabbed megamenus can have lots of content in them, creating a large DOM footprint and consequently a relatively heavy tax on device memory. This PR reduces this footprint by only rendering the active tab's tabpanel.

To facilitate this, we needed to refactor the way we "lazy-load" the megamenus so that they make use of Lit's rendering lifecycle. Previously they injected markup straight into the DOM, but this markup was somehow disconnected from Lit's reactivity mechanisms. As a result, updating a state property didn't successfully update the DOM content of the rendered megamenu.

### Changelog

**New**

- Only render the active tab's tabpanel in tabbed megamenus.

**Changed**

- Refactors masthead lazy-loading to use Lit's rendering lifecycle.

### To Test

Visit the [deploy preview](https://ibmdotcom-webcomponents.s3.us-east.cloud-object-storage.appdomain.cloud/deploy-previews/10815/index.html?path=/story/components-masthead--default) and verify the following with regards to tabbed megamenus:
- They render just the same as [the latest released version](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/masthead-v2/web-components/?path=/story/components-masthead--default).
- Only the active tab's tabpanel (i.e. the right side of the megamenu) is present in the DOM.

### CODE REVIEW

File: packages/web-components/src/components/masthead/megamenu-tabs.ts [hunk 1]
```diff
@@ -35,6 +35,14 @@ class DDSMegaMenuTabs extends BXTabs {
     return `${ddsPrefix}-megamenu-tab[selected]`;
   }
 
+  /**
+   * The name of the custom event fired before a tab is selected upon a user gesture.
+   * Cancellation of this event stops changing the user-initiated selection.
+   */
+  static get eventBeforeSelect() {
+    return `${ddsPrefix}-megamenu-tabs-beingselected`;
```
Comment (andy-blum) [inline packages/web-components/src/components/masthead/megamenu-tabs.ts:43]: We're defining this same event name in masthead-composite.ts:1554. Should we really have the same event title defined on two different components? Are they both in use?

**Review by andy-blum — COMMENTED** [review]
The one e2e test that failed did so _after_ all cypress tests were marked as successful, and a manual testing of the dds-top-nav and megamenu panels appears to be working well, even with CPU and network throttling.

One tiny question in the code, but not a blocker.

**Review by andy-blum — APPROVED** [review]

File: packages/web-components/src/components/masthead/megamenu-tabs.ts (hunk 1 above)
Comment (jkaeser) [reply packages/web-components/src/components/masthead/megamenu-tabs.ts:43]: This is the library's existing pattern of emitting and listening for custom events. Here's another example used in the Masthead:
- Emitting component: https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/feat/masthead-v2-dev/packages/web-components/src/components/masthead/megamenu-top-nav-menu.ts#L94
- Listening component: https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/feat/masthead-v2-dev/packages/web-components/src/components/masthead/masthead-composite.ts#L1500-L1502

Now that you call attention to it, though, I don't like the pattern very much. We have to use the same magic string twice, whereas we could just as easily define it as a static property on the emitting class, then access that property on the listening class. I'll update.

File: packages/web-components/src/components/masthead/masthead-composite.ts [hunk 2]
```diff
@@ -967,24 +981,52 @@ class DDSMastheadComposite extends HostListenerMixin(LitElement) {
...
   protected _loadMegamenu = (event: CustomEvent) => {
     const {
       target,
       detail: { active, resolveFn },
     } = event;
     const { autoid } = (target as HTMLElement).dataset;
-    const index = autoid?.slice(-1);
-    const currentMenu = this.megamenuSet[index!];
-    render(active ? currentMenu : nothing, target as HTMLElement);
+    const targetMegamenuIndex = Number(autoid?.slice(-1));
```
Comment (m4olivei) [inline packages/web-components/src/components/masthead/masthead-composite.ts:996]: Not a regression, but just noticing the `autoid?.slice(-1)` is fragile for the case when there are more than 9 items. Lets say there were 11, the first item and 11th item would both produce a `targetMegamenuIndex` of `1`. Very edge case, and not necessary to handle here, but I couldn't unsee it.

**Review by m4olivei — APPROVED** [review]
LGTM. Just minor notes, not blocking. Nice work on this!

File: packages/web-components/tests/e2e-storybook/cypress/integration/masthead/masthead-l1.e2e.js [hunk 3]
```diff
@@ -74,7 +74,7 @@ describe('dds-masthead | with L1 (desktop)', () => {
       .find(`${_selectors.l1Item}`)
       .first()
       .next(_selectors.l1Dropdown)
-      .should('be.visible');
+      .should('have.length', 1);
```
[the remark this answers is not in the document]
Comment (jkaeser) [reply packages/web-components/tests/e2e-storybook/cypress/integration/masthead/masthead-l1.e2e.js]: Just getting tests to pass. Updated to match the work you did in https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/10867/commits/616f7f636c53bd98b32ca8f1b3296b02a49f0135

**Comment by m4olivei** [comment]
Thanks for those changes @jkaeser . They look great!
